### ISSUE TITLE
Fix `at-rule-prelude-no-invalid` `Bad syntax reference` throw for `@when` and `@else`

### ISSUE THREAD

**Issue by elvism-lullabot** [stylelint/stylelint#9511]
### What minimal example or steps are needed to reproduce the bug?

I have the following SCSS:
```scss
  @if $min and $max {
    @media (min-width: $min) and (max-width: $max) { @content; }
  } @else if $min {
    @media (min-width: $min) { @content; }
  } @else if $max {
    @media (max-width: $max) { @content; }
  } @else {
    @error 'media-query() needs $from and/or $until.';
  }
```
After updating to `stylelint:^17.15.0`, I'm seeing the following error when running `yarn stylelint "**/*.{css,scss,sass}"`
```
yarn stylelint "**/*.{css,scss,sass}"
yarn run v1.22.22
warning ../../../package.json: No license field
$ stylelint '**/*.{css,scss,sass}'
Error: Bad syntax reference: <boolean-condition>
    at /home/developer/project/test.scss:3:3
    at internalMatch (file:///home/developer/project/node_modules/css-tree/lib/lexer/match.js:406:27)
    at matchAsTree (file:///home/developer/project/node_modules/css-tree/lib/lexer/match.js:586:25)
    at matchSyntax (file:///home/developer/project/node_modules/css-tree/lib/lexer/Lexer.js:88:18)
    at Lexer.matchAtrulePrelude (file:///home/developer/project/node_modules/css-tree/lib/lexer/Lexer.js:337:16)
    at file:///home/developer/project/node_modules/stylelint/lib/rules/at-rule-prelude-no-invalid/index.mjs:49:28
    at /home/developer/project/node_modules/stylelint/node_modules/postcss/lib/container.js:377:18
    at Root.walk (/home/developer/project/node_modules/stylelint/node_modules/postcss/lib/container.js:352:18)
    at Root.walkAtRules (/home/developer/project/node_modules/stylelint/node_modules/postcss/lib/container.js:375:19)
    at file:///home/developer/project/node_modules/stylelint/lib/rules/at-rule-prelude-no-invalid/index.mjs:42:8
    at runRule (file:///home/developer/project/node_modules/stylelint/lib/lintPostcssResult.mjs:140:11)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Minimal example:**
```scss
@if true {
  @media (min-width: 1em) and (max-width: 2em) { color: red; }
} @else if true {
  @media (min-width: $min) { color: red; }
}
```

[Demo link](https://stylelint.io/demo/#N4Igxg9gJgpiBcIACBLAZgAgC4CcCuMGwAOgHYYZIC2MUKAhhgBRUqkC0A7ilFgBbwMARhhUAlBnqkozKvQAeXHv0EAmUROAZIAGwg5BOWgG4MAXzJnKMHQGdC6bPkIlylGnUYs2S3gIwAJKykmtoQegYYRlCmFqRmIAA0IGgoOjAAcvQ0CCAw8tkADukAdGC2tkngEKSpAOa5+YX6WBiwaPR4Oq2uFPlYMNK2ggDaxCC2WACe6TpsWOyQtSh17JNSUPQ4UOOJGOOTMzbzizX1a1gbW1Br5bbjALqJZBRgeJMQVADKU6SX8oJxs1JndbhVxpZjGQqkt6gAxfRyLC5Q6zeZlM4rEpUABWlWSsEKlUQvX2E2maL+40BIB09AGk12LzJqOOf1Oy1W62k12pZIAzAAOEoABlFTLcBwpbIWsJWFyu2zB9wQZKEABZReKksypUc5uzbHc+eMAOwlVTayzQsxAA)

### What minimal configuration is needed to reproduce the bug?

```json
{
  "extends": ["stylelint-config-standard", "stylelint-config-standard-scss"],
  "customSyntax": "postcss-scss"
}
```

### How did you run Stylelint?

```
yarn stylelint "**/*.{css,scss,sass}"
```

### Which Stylelint-related dependencies are you using?

```json
{
  "stylelint": "^17.15.0",
  "stylelint-config-standard": "^38.0.0",
  "stylelint-config-standard-scss": "^14.0.0",
  "stylelint-scss": "^7.2.0"
}
```

### What did you expect to happen?

No errors

### What actually happened?

Error: Bad syntax reference: `<boolean-condition>`

**Comment by jeddy3** [comment]
@elvism-lullabot Thanks for the clear report and for using the template.

It looks like a bug in the [`at-rule-prelude-no-invalid`](https://stylelint.io/user-guide/rules/at-rule-prelude-no-invalid/) rule when it attempts to lint the `@when` and `@else` at-rules from the [CSS Conditional Rules Module Level 5](https://www.w3.org/TR/css-conditional-5/) specification. Triggered by an update to our CSSTree patches that [now includes their preludes](https://github.com/csstools/postcss-plugins/blob/7d2aa1364a47b0be14330b7fbceb51f0d2343ef8/packages/css-syntax-patches-for-csstree/dist/index.json#L24-L26):

```json
{
  "else": {
    "prelude": "<boolean-condition>?"
  }
}
```

Reproducible in the [Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIACB3AFjAdgAgLYygEsBDAChUKgBc1sA+AXmwBYAGVgBwA8BKbYAL7YAOpiQwANgGcY2KQFcOHCACcqU0mGIqYVALSQJq+Ng6FMAaz6CQAGhAAzQhJgA5YvgQgYXDxxcAdGBSUnbgEJhOAOZePspq2LAOxPISVPyi2NgqqTBSJsCZWSIgxPo5LnocOhLysHqYEHrmAG7EEpTCCNhUOTBFAqICANyiYZCRhFEAYqq4ZV5SVACeLh2YVEER0QG4AFah9rAcoYiFOCVLq5LmVF0mXRJleXd2RV1Xa7cG21N6S8RMFBtFB7iUAMwADgCrBhXVs7xAnxuGx+kyi-yogOBKig-2CUjBXQAjMwYXC3hcPisvqipASiSAAOwBABMFKGYwEQA) with:

```css
@when media(width >= 400px) {} 
@else supports(caret-color: pink) {}
```

And:

```js
export default {
  rules: {
    "at-rule-prelude-no-invalid": true
  }
};

```

However, you shouldn't have encountered the bug, as you're linting SCSS rather than [upcoming CSS](https://caniuse.com/css-when-else).

You can avoid it by updating to the latest version of `stylelint-config-standard-scss`, which [turns off the `at-rule-prelude-no-invalid` rule](https://github.com/stylelint-scss/stylelint-config-recommended-scss/blob/9492951fbaaf78558b215550fe2707b63e8f9f5e/index.js#L12). The version you're using, `14.0.0`, is two years old. 

As you seem to be linting both CSS and SCSS, I recommend using the [`overrides`](https://stylelint.io/user-guide/configure/#overrides) configuration property, so that your:

- CSS is linted with `stylelint-config-standard`
- SCSS is linted with `stylelint-config-standard-scss`

That way, rules like `at-rule-prelude-no-invalid` will be enabled for CSS and disabled for SCSS.

```json
{
  "extends": ["stylelint-config-standard"],
   "overrides": [
    {
      "files": ["*.scss", "**/*.scss"],
      "extends": ["stylelint-config-standard-scss"],
    }
  ]
}
```

You also don't need to manually include `postcss-scss` as it's bundled and configured by `stylelint-config-standard-scss`.

I've labelled the `@when`/`@else` spec issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/contributor-guide/rules#fix-a-bug-in-a-rule) in the Contributor guide.

**Comment by romainmenke** [comment]
Thank you for the detailed analysis @jeddy3 🙇 
I have excluded `when` and `else` from the syntax patches for now.
So the demo you linked is already free of errors :)

**Comment by jeddy3** [comment]
> Thank you for the detailed analysis @jeddy3 🙇

No problemo 😁 .

I find that these types of bugs are always a good opportunity to learn about upcoming additions to the language.

>  I have excluded when and else from the syntax patches for now.

Fantastic, thank you.

I'll leave the issue as ready to implement in case anyone else has time to dig into the underlying cause, here or upstream in patches or the spec, so we can reinclude `when` and `else`.

**Comment by Abdellox** [comment]
Thanks for opening this! I'd like to help investigate.

Could you share:
- Steps to reproduce
- Expected vs actual behavior
- Environment details

I'll take a look at the code and see if I can help.
